packages/flutter/lib/src/services/deferred_component.dart

```diff
@@ -0,0 +1,157 @@
+/// be repeated to trigger the loading process.
+enum DeferredComponentInstallState {
+  /// The deferred component installation has been requested, but has not begun
+  /// download or installation yet.
+  requested,
+  /// The deferred component installation is awaiting further action.
+  pending,
+  /// User input is requried for continued installation.
+  requireUserConfirmation,
+  /// The deferred component is being downloaded.
+  downloading,
+  /// The deferred component has finished downloading but has not yet been
+  /// installed.
+  downloaded,
+  /// The downloaded deferred component is being installed.
+  installing,
+  /// The deferred component has previously been installed and all files/assets
+  /// are available, but has not been loaded into the VM in the current
+  /// app's session. Either `loadLibrary()` or [DeferredComponent.installDeferredComponent]
+  /// should be called to trigger the loading process to ensure the deferred component's
+  /// components can be safely used in the app.
+  installedPendingLoad,
+  /// The deferred component has been succesfully downloaded and installed. Assets and
+  /// code from the deferred component are ready to use.
+  installed,
+  /// An installation request has been asked to be canceled.
+  cancelling,
+  /// An installation request has been canceled.
+  canceled,
+  /// An installation request has failed.
+  failed,
+  /// The deferred component has not been requested nor installed or is unavailable.
+  unknown,
+}
+
+/// Manages the installation and loading of deferred component modules.
+///
+/// Deferred components allow Flutter applications to download precompiled AOT
+/// dart code and assets at runtime, reducing the install size of apps and
+/// avoiding installing unnessecary code/assets on end user devices. Common
+/// use cases include deferring installation of advanced or infrequently
+/// used features and limiting locale specific features to users of matching
+/// locales.
+class DeferredComponent {
+  // This class is not meant to be instantiated or extended; this constructor
+  // prevents instantiation and extension.
+  // ignore: unused_element
+  DeferredComponent._();
+
+  // TODO(garyq): We should eventually expand this to install modules by loadingUnitId
+  // as well as moduleName, but currently, loadingUnitId is opaque to the dart code
+  // so this is not possible. The API has been left flexible to allow adding
+  // loadingUnitId as a parameter.
+
+  /// Requests that an assets-only deferred component identified by the [moduleName]
+  /// be downloaded and installed.
+  ///
+  /// This method returns a Future<void> that will complete when the feature is
+  /// installed and any assets are ready to used. When an error occurs, the
+  /// future will complete an error.
+  ///
+  /// This method should be used for asset-only deferred components. Deferred components
+  /// containing dart code should call `loadLibrary()` on a deferred imported
```

Comment:
"should call" or "must call"? i.e. is there every a reason to use this on a dart code component?